src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineView.xaml.cs

```diff
@@ -0,0 +1,183 @@
+            => EnqueueFilterAndSortTask();
+
+        private void SortByName(object sender, EventArgs e)
+            => SetSortOptionAndUpdateDataModel(SortOption.Name);
+
+        private void SortByOrder(object sender, EventArgs e)
+            => SetSortOptionAndUpdateDataModel(SortOption.Location);
+
+        private void SortByType(object sender, EventArgs e)
+            => SetSortOptionAndUpdateDataModel(SortOption.Type);
+
+        private void SetSortOptionAndUpdateDataModel(SortOption sortOption)
+        {
+            _viewModel.SortOption = sortOption;
+            EnqueueFilterAndSortTask();
+        }
+
+        private void EnqueueFilterAndSortTask()
+        {
+            var service = _visualStudioCodeWindowInfoService.GetServiceAndThrowIfNotOnUIThread();
+            var caretPoint = service.GetCurrentCaretSnapshotPoint();
+            _viewModel.EnqueueFilterAndSortTask(caretPoint);
+        }
+
+        private void EnqueueUIUpdateTask(ExpansionOption option)
+        {
+            var service = _visualStudioCodeWindowInfoService.GetServiceAndThrowIfNotOnUIThread();
+            var caretPoint = service.GetCurrentCaretSnapshotPoint();
+            _viewModel.EnqueueUIUpdateTask(option, caretPoint);
+        }
+
+        /// <summary>
+        /// When a symbol node in the window is clicked, move the caret to its position in the latest active text view.
+        /// </summary>
+        private void SymbolTree_MouseDown(object sender, EventArgs e)
+        {
+            if (sender is StackPanel panel && panel.DataContext is DocumentSymbolUIItem symbol)
+            {
+                var service = _visualStudioCodeWindowInfoService.GetServiceAndThrowIfNotOnUIThread();
+                var activeTextView = service.GetLastActiveIWpfTextView();
+                if (activeTextView is null)
+                    return;
+
+                // When the user clicks on a symbol node in the window, we want to move the cursor to that line in the editor. If we
+                // don't unsubscribe from Caret_PositionChanged first, we will call EnqueueHighlightExpandAndPresentItemsTask() once
+                // we move the cursor ourselves. This is not ideal because we would be doing extra work to update the view with an
+                // identical document symbol tree.
+                activeTextView.Caret.PositionChanged -= Caret_PositionChanged;
+
+                // Prevents us from being permanently unsubscribed if an exception is thrown while updating the text view selection.
+                try
+                {
+                    // Map the symbol's selection range start SnapshotPoint to a SnapshotPoint in the current textview then set the
+                    // active text view caret position to this SnapshotPoint.
+                    activeTextView.TryMoveCaretToAndEnsureVisible(
+                        symbol.SelectionRangeSpan.Start.TranslateTo(activeTextView.TextSnapshot, PointTrackingMode.Negative));
```

Comment:
this seems wrong.  we shouldn't use the text-snapshot of the text-view.  It might not actually be a roslyn snapshot (like with projection buffers).

Reply:
We always get this from the codewindow which is guranteed to be applicable to roslyn.

Reply:
can you link me to taht code?